### DISCUSSION TITLE
No transactions when running tests on anvil

### DISCUSSION THREAD

**Discussion by Keinberger** [foundry-rs/foundry#5421] [Help] [answered]
### Hey there!

I have an issue/question about running local tests on an anvil node. When I spin up a local anvil node and try to run my foundry tests there, I only see a couple of calls being made to the anvil node, however, there are **no transactions being published or blocks being mined**. If I try to deploy a contract to the node, I can successfully see the transaction being published, but when I run my foundry tests (which use foundyry cheatcodes), only RPC calls show up.

My question: Is that expected behaviour? Are foundry tests when specifying `--rpc-url` not actually being run on the blockchain, but rather the blockchain state is being copied and reproduced again for the testing environment?

I wanted to have a ethers/viem listener running to capture the transactions of my tests, which should run on the anvil node, but it seems like something like that is not possible?

**Comment by aathan** [comment, marked as answer]
AFAIK `foundry test` does not execute its tests in the node, but rather, within its own local evm. I.e., `foundry test` is not "deploying" the test contracts themselves into the node. Instead it is using the node to get blockchain state from which to create an internal execution context to run the test. I'm still studying all the foundry source code so take what I just said with that in mind. Arguably, at some point in the future, foundry test could work by running the tests in the node, checkpointing & reverting the chain prior to each run of the test. I'm not clear on how much (if any) code duplication exists since both environments need to handle cheatcodes etc, but they probably share the "revm" implementation. I could see running tests in the node being useful if there are other "attachments" that could be made to the anvil node, including longer lived blockchain state persistence or test runner/etc stuff that you may want to survive across test runs...

**Reply by Keinberger** [reply]
Hey, thanks for your answer. So it seems like that you have validated my theory of Foundry currently  not being able to support running tests on Anvil directly as of now.

Would be interesting to see if there are any active discussion to support Anvil as a testing node. That way external sources could be integrated with the tests, which could be highly usable sometimes.

**Reply by aathan** [reply]
I am certain that anvil supports console/console2 callas at the hardhat console address, and AFAIK anvil itself *does* support cheatcodes in general, since all of foundry uses the same evm implementation -- though I have not confirmed the latter statement. So, "running tests on Anvil" is an human language statement that needs to be made more concrete in order to answer your question.

Can you use hardhat.js or foundry scripts to deploy contracts to an anvil fork, and then run them there? Yes.
Can those contracts contain console.log() calls? Yes.
Can those contracts contain prank() etc calls? I think so.
Can those contracts be deployed and invoked using the `foundry test` command? No.
Can those contracts be "tests" in some sense of the word? Yes.

Even if the cheat codes are not available, it's common practice to do things like fork a main chain, and then interact with uniswap pools and other "well known" contracts on the chain, but inside the fork, thereby testing your code. This could include creating fake ERC20 coins with which to instantiate new pools, or replacing the contract code of a major ERC20 like USDC with your own (using a cheatcode).

**Reply by Ibby-debug** [reply]
I'm curious to know how would I be able to run hardhat runner scripts against a forge deployed contract? I wanna run like node.js a process that interacts with a deployed contract.

**Comment by UdeRox** [comment]
> ### 
> 
> My question: Is that expected behaviour? Are foundry tests when specifying `--rpc-url` not actually being run on the blockchain, but rather the blockchain state is being copied and reproduced again for the testing environment?
> 

when you specifying --rpc-url, that means you are asking to run the test on node, you can get the rpc-url from alchemy.com for different network, you can see the transactions in the dashboard in the alchemy.com. 
when you run without --rpc-url, yes then test runs in the local environment, to run test we might need some on chain contracts, to resolve on chain contracts, as strategy we use mocking, in this case mock contract will deploy anvil local net work, this is explained well in the video clearly

**Reply by aathan** [reply]
I don't use alchemy.com so I am not sure exactly what transactions you're saying you see there. However, `foundry test` for sure does not end up executing tests on-chain. The rpc-url is used to issue eth rpc calls to get blockchain state, and this state is used by the internal evm to run the tests. Perhaps you mean that in the alchemy.com dashboard, you see the results of those calls. That is, however, not the same thing as having your tests cause on-chain storage to change. If that were true, you would not be able to run the same test twice without re-forking from a block prior to any changes your calls make to external contracts ... which is not necessary, therefore what I'm saying is correct.

Happy to understand what you mean better if you disagree with me.

**Reply by Keinberger** [reply]
Can only iterate what @aathan answered on this. RPC URL's are used to issue RPC calls, to retrieve information of the blockchain state or publish data to the blockchain. In the case of running tests with Foundry, Foundry only retrieves information of the specified RPC endpoint and copies that state to the local testing node. So there are no transactions actually getting published on the node specified in the RPC URL.

**Reply by UdeRox** [reply]
@aathan @Keinberger yes, I agree with you my understanding was wrong :)
